### PR TITLE
Add target-shell's ls command options

### PR DESCRIPTION
**Pull request by mnrkbys**
Added following target-shell's ls command options:
```
  -u                    Show time of last access.
  -c                    Show time when file status was last changed.
  -R                    Recursively list subdirectories encountered.
```

### CODE REVIEW

**Comment by pyrco** [comment]
Thanks for these useful additions! Once the CLA accepted, we will plan reviewing this PR.

**Comment by mnrkbys** [comment]
@DissectBot agree

File: dissect/target/tools/shell.py [hunk 1]
```diff
@@ -504,20 +512,61 @@ def cmd_ls(self, args: argparse.Namespace, stdout) -> Optional[bool]:
...
+            for target_path, name in contents:
+                self.print_extensive_file_stat(args=args, stdout=stdout, target_path=target_path, name=name)
+                if args.recursive and target_path.is_dir():
+                    dir_list.append(target_path)
+
+        if args.recursive and len(dir_list):
+            for dir_name in dir_list:
+                self._recursive_ls(args, dir_name, stdout)
+
+    def _recursive_ls(self, args: argparse.Namespace, dir_name: fsutil.TargetPath, stdout: TextIO) -> Optional[bool]:
```
Comment (Schamper) [inline dissect/target/tools/shell.py:531]: There's currently quite a bit of code duplication between this and `cmd_ls`. Do you think you could re-use more? I.e. let `_recursive_ls` be the main body (perhaps rename it something like `_print_ls`?), and do the print on line 540 conditionally if the function is being called recursively (through an additional argument).

File: dissect/target/tools/shell.py [hunk 2]
```diff
@@ -504,20 +512,61 @@ def cmd_ls(self, args: argparse.Namespace, stdout) -> Optional[bool]:
...
             entry = target_path.get()
             stat = entry.lstat()
             symlink = f" -> {entry.readlink()}" if entry.is_symlink() else ""
-            utc_time = datetime.datetime.utcfromtimestamp(stat.st_mtime).isoformat()
+            if args.use_ctime:
+                utc_time = datetime.datetime.utcfromtimestamp(stat.st_ctime).isoformat()
+            elif args.use_atime:
+                utc_time = datetime.datetime.utcfromtimestamp(stat.st_atime).isoformat()
+            else:
+                utc_time = datetime.datetime.utcfromtimestamp(stat.st_mtime).isoformat()
```
Comment (Schamper) [inline dissect/target/tools/shell.py:569]:
```suggestion
            show_time = stat.st_mtime
            if args.use_ctime:
                show_time = stat.st_ctime
            elif args.use_atime:
                show_time = stat.st_atime
            utc_time = datetime.datetime.utcfromtimestamp(show_time).isoformat()
```

File: dissect/target/tools/shell.py [hunk 3]
```diff
@@ -490,10 +490,18 @@ def do_info(self, line: str) -> Optional[bool]:
     @arg("-l", action="store_true")
     @arg("-a", "--all", action="store_true")  # ignored but included for proper argument parsing
     @arg("-h", "--human-readable", action="store_true")
-    def cmd_ls(self, args: argparse.Namespace, stdout) -> Optional[bool]:
+    @arg("-R", action="store_true", dest="recursive", help="Recursively list subdirectories encountered.")
+    @arg("-c", action="store_true", dest="use_ctime", help="Show time when file status was last changed.")
+    @arg("-u", action="store_true", dest="use_atime", help="Show time of last access.")
```
Comment (Schamper) [inline dissect/target/tools/shell.py:495]:
```suggestion
    @arg("-R", "--recursive", action="store_true", help="recursively list subdirectories encountered")
    @arg("-c", action="store_true", dest="use_ctime", help="show time when file status was last changed")
    @arg("-u", action="store_true", dest="use_atime", help="show time of last access")
```
The lowercasing is mostly for consistency with an average bash shell.

File: dissect/target/tools/shell.py [hunk 4]
```diff
@@ -490,10 +490,18 @@ def do_info(self, line: str) -> Optional[bool]:
...
+    def cmd_ls(self, args: argparse.Namespace, stdout: TextIO) -> Optional[bool]:
         """list directory contents"""
 
         path = self.resolve_path(args.path)
+        dir_list = []
+
+        if args.use_ctime and args.use_atime:
+            print("Cannot specify -c and -u at the same time.")
```
Comment (Schamper) [inline dissect/target/tools/shell.py:503]:
```suggestion
            print("can't specify -c and -u at the same time")
```
The lowercasing is mostly for consistency with an average bash shell.

File: dissect/target/tools/shell.py [hunk 5]
```diff
@@ -504,20 +512,61 @@ def cmd_ls(self, args: argparse.Namespace, stdout) -> Optional[bool]:
...
+    def _recursive_ls(self, args: argparse.Namespace, dir_name: fsutil.TargetPath, stdout: TextIO) -> Optional[bool]:
+        path = self.resolve_path(dir_name)
+        dir_list = []
+
+        if path.is_dir():
+            contents = self.scandir(path, color=True)
+        elif path.is_file():
+            contents = [(path, path.name)]
+
+        print(f"\n{str(path)}:")
```
Comment (Schamper) [inline dissect/target/tools/shell.py:540]:
```suggestion
        print(f"\n{str(path)}:", file=stdout)
```

**Review by Schamper — CHANGES_REQUESTED** [review]

**Comment by mnrkbys** [comment]
I was very busy and could not fix it for a while.
Now, I have modified the code according to the comments.

File: dissect/target/tools/shell.py [hunk 6]
```diff
@@ -490,34 +490,66 @@ def do_info(self, line: str) -> Optional[bool]:
...
                 print(f"total {len(contents)}", file=stdout)
             for target_path, name in contents:
-                self.print_extensive_file_stat(stdout=stdout, target_path=target_path, name=name)
+                self.print_extensive_file_stat(args=args, stdout=stdout, target_path=target_path, name=name)
+                if target_path.is_dir():
+                    dir_list.append(target_path)
+
+        if args.recursive and len(dir_list):
+            for dir_name in dir_list:
+                self._print_ls(args, dir_name, dir_depth+1, stdout)
```
Comment (Schamper) [inline dissect/target/tools/shell.py:539]:
```suggestion
                self.print_extensive_file_stat(args=args, stdout=stdout, target_path=target_path, name=name)
                if target_path.is_dir():
                    subdirs.append(target_path)

        if args.recursive and subdirs:
            for subdir in subdirs:
                self._print_ls(args, subdir, depth + 1, stdout)
```

File: dissect/target/tools/shell.py [hunk 7]
```diff
@@ -490,34 +490,66 @@ def do_info(self, line: str) -> Optional[bool]:
...
 
+        if dir_depth > 0:
+            print(f"\n{str(path)}:", file=stdout)
+
         if not args.l:
-            print("\n".join([name for _, name in contents]), file=stdout)
+            for target_path, name in contents:
+                print(name, file=stdout)
+                if target_path.is_dir():
+                    dir_list.append(target_path)
```
Comment (Schamper) [inline dissect/target/tools/shell.py:528]:
```suggestion
                    subdirs.append(target_path)
```

**Review by Schamper — CHANGES_REQUESTED** [review]
Thanks! Just some small stylistic suggestions. Could you also rebase on the latest main branch?

**Comment by mnrkbys** [comment]
Thanks for your comment.
I have updated shell.py again and rebased on the latest main.

**Pull request merged**